### PR TITLE
core: Optimize *.WithDirectory and *.WithFile via MergeOp.

### PR DESCRIPTION
**Pull request by sipsma**
We now use MergeOp as much as possible to optimize copying from one LLB state onto another. See the comments in the mergeStates function in core/directory.go for more details.

### CODE REVIEW

**Review by gerhard — APPROVED** [review]
LGTM 💪

File: core/directory.go [hunk 1]
```diff
@@ -316,30 +316,117 @@ func (dir *Directory) WithDirectory(ctx context.Context, subdir string, src *Dir
...
+		mergeStates = append(mergeStates, input.Src)
+	} else {
+		// Even if we can't merge directly, we can still get some optimization by
+		// copying to scratch and then merging that. This still results in an on-disk
+		// copy but preserves the other caching benefits of MergeOp. This is the same
+		// behavior as "COPY --link" in Dockerfiles.
+		mergeStates = append(mergeStates, llb.Scratch().File(llb.Copy(
+			input.Src, path.Join(input.SrcDir, input.SrcFileName), path.Join(input.DestDir, input.DestFileName), copyInfo,
+		)))
+	}
```
Comment (vito) [inline core/directory.go:426]: I remember being tempted by Merge here but realizing it's more complicated because of the various configs, so it's really cool to see both situations optimized!

File: core/integration/directory_test.go [hunk 2]
```diff
@@ -821,3 +823,50 @@ func TestDirectoryWithFileExceedingLength(t *testing.T) {
...
+			WithExec([]string{"sh", "-e", "-x", "-c",
+				"mkdir -p " + parentDir + " && touch " + filePath + " && stat -c '%i' " + filePath,
+			})
+		out, err := ctr.Stdout(ctx)
+		require.NoError(t, err)
+		return ctr.Directory(parentDir), strings.TrimSpace(out)
+	}
+
+	// verify optimized hardlink based merge-op is used by verifying inodes are preserved
+	// across WithDirectory calls
```
Comment (vito) [inline core/integration/directory_test.go:848]: neat :nerd_face:

File: cmd/shim/main.go [hunk 3]
```diff
@@ -250,6 +250,9 @@ func shim() int {
 		}
 
 		code, err := os.ReadFile(exitCodePath)
+		if os.IsNotExist(err) {
+			return -1
+		}
```
Comment (vito) [inline cmd/shim/main.go:255]: Thanks for fixing this along the way, I've seen it pop up here and there but it was always some tertiary failure so I never doubled back.

**Review by vito — APPROVED** [review]
lgtm!

File: cmd/shim/main.go (hunk 3 above)
Comment (sipsma) [reply cmd/shim/main.go:255]: Oh yeah no problem, I had accidentally mounted something at `/` and was very confused for a bit... 😄

**Comment by sipsma** [comment]
I was about to merge but looked at the code one more time and noticed a corner cases where if you use `WithFile` and the full path to the file is the same is dest+source, you would include both the file and any other files in it's parent dir.

The fix was very trivial, added test coverage too: https://github.com/dagger/dagger/pull/5400/commits/42dd2eb73b530b43ae74ac0e32d814a7b7ab56e6

Given it was a teeny modification I'll still go ahead and merge this once CI is done
